### PR TITLE
Add join='override'

### LINKED ISSUE #2023: quantile method returns quantile coordinates which can raise issues
I browsed the issues to check whether there was an open issue on this, but didn't find anything.
Apologies if it is a duplicate.

I am an happy user of the new quantile method in xarray.  I think though that creating a coordinate that has the same name as the method can raise issues.

While updating my code to switch entirely to xarray and the quantile method, I realized that this could be an issue. In the old code I had some datasets created using np.nanpercentiles, in which I had generated the coordinate and named it "q". 
At some point in my code I have to retrieve the list of the q values, I used to do that by extracting the values:

```python
x = ds.q.values
```
updating it to what would be the coordinate automatically generated by xarray, it should be

```python
x = ds.quantile.values
```

#### Problem description

But of course, when trying to call the coordinate "quantile" from the dataset, you get an Attribute Error:

```python
AttributeError: 'function' object has no attribute 'values'
```
This can be easily solved by accessing the coordinate as:

```python
x = ds['quantile'].values
```

However, I thought to point out that using the same name for the coordinate and the method can raise this type of issues.

### PR DESCRIPTION
**Pull request by dcherian**
This adds `join='override'` which checks that indexes along a dimension are of the same size and overwrites those indices with indices from the first object.

Definitely need help, feedback.

 - [x] With #3102 this partially helps with #2039, #2217
 - [x] Tests added
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

File: xarray/core/dataarray.py [hunk 1]
```diff
@@ -329,7 +329,7 @@ def _replace_maybe_drop_dims(
                                  if set(v.dims) <= allowed_dims)
         return self._replace(variable, coords, name)
 
-    def _replace_indexes(
+    def _overwrite_indexes(
```
Comment (dcherian) [inline xarray/core/dataarray.py]: This wasn't used anywhere in the codebase so I renamed it to be consistent with `Dataset._overwrite_indexes`.

File: xarray/core/alignment.py [hunk 2]
```diff
@@ -102,6 +108,20 @@ def align(*objects, join='inner', copy=True, indexes=None, exclude=frozenset(),
...
+        for index, obj in enumerate(objects[1:]):
+            for dim in objects[0].indexes:
+                if dim not in exclude:
+                    if obj.indexes[dim].size != objects[0].indexes[dim].size:
+                        raise ValueError("Cannot use join='override' when "
+                                         "all indexes don't have the same "
+                                         "shape.")
+                    new_indexes[dim] = objects[0].indexes[dim]
+
+            objects[index + 1] = obj._overwrite_indexes(new_indexes)
```
Comment (dcherian) [inline xarray/core/alignment.py]: Is this the right way to do this?

**Comment by crusaderky** [comment]
Hi,

This doesn't close #2039 because

- open_mfdataset does not offer a 'join' parameter to be propagated to align
- even if it did, by when open_mfdataset invokes align time has already been wasted because indices, as of today, are eagerly loaded by open_dataset
- it does nothing about non-index variables whose dims don't include the concat_dim

**Comment by dcherian** [comment]
Ah, good point. Do you have any thoughts on the implementation? or on the `compat='override'` proposal in https://github.com/pydata/xarray/issues/2064#issuecomment-519149757

**Comment by crusaderky** [comment]
Lazy-loading indices, plus this and #3102, would probably solve #2039.
Without lazy-loading, an ad-hoc needs to be implemented in open_mfdataset directly.

**Comment by dcherian** [comment]
Ready for review.

File: xarray/core/alignment.py (hunk 2 above)
Comment (shoyer) [reply xarray/core/alignment.py]: It looks like this should work, yes.

File: xarray/core/alignment.py [hunk 3]
```diff
@@ -99,6 +105,21 @@ def align(
         obj, = objects
         return (obj.copy(deep=copy),)
 
+    if join == "override":
+        objects = list(objects)
+        new_indexes = dict()
+        for index, obj in enumerate(objects[1:]):
+            for dim in objects[0].indexes:
```
Comment (shoyer) [inline xarray/core/alignment.py]: There is a potential edge case here, where the first object does not have one of the dimensions found on multiple later objects, e.g.,
`align(DataArray(0), DataArray([1], [('x', [1])]), DataArray([2, 3], [('x', [2, 3])])`

This should be an error, but I think the current code would let this pass through, possibly creating an invalid object. Instead, I think the right strategy is to iterate everything, building up the set of indexes to match against as you go.

Another edge case that I think you might get wrong currently is something like this, where the last DataArray doesn't have one of the indexes at all:
`align(DataArray([1], [('x', [1])]), DataArray([2], [('x', [2])], DataArray(3))`

File: xarray/core/alignment.py [hunk 4]
```diff
@@ -99,6 +105,21 @@ def align(
         obj, = objects
         return (obj.copy(deep=copy),)
 
+    if join == "override":
```
Comment (shoyer) [inline xarray/core/alignment.py]: Could the logic from this block be moved to a helper function? `align` is already getting pretty long.

File: xarray/core/alignment.py (hunk 3 above)
Comment (dcherian) [reply xarray/core/alignment.py]: Ah thanks. Fixed now. I've added these two examples as a test case.

File: xarray/core/alignment.py [hunk 5]
```diff
@@ -31,10 +31,34 @@ def _get_joiner(join):
...
+        # We rewrite all indexes and then use join='left'
+        return operator.itemgetter(0)
     else:
         raise ValueError("invalid value for join: %s" % join)
 
 
+def _override_indexes(objects, all_indexes, exclude):
+    for dim, dim_indexes in all_indexes.items():
+        if dim not in exclude:
+            lengths = [dd.size for dd in dim_indexes]
```
Comment (shoyer) [inline xarray/core/alignment.py]: Could we just use a set comprehension here? Also I have a minor distaste for abbreviations like `dd`.

e.g., maybe
``python
            lengths = {index.size for index in dim_indexes}
            if len(lengths) != 1:
                raise ValueError(...)
```

File: xarray/core/alignment.py [hunk 6]
```diff
@@ -31,10 +31,34 @@ def _get_joiner(join):
...
+            if not all([ll == lengths[0] for ll in lengths[1:]]):
+                raise ValueError(
+                    "Indexes along dimension %r are not equal."
```
Comment (shoyer) [inline xarray/core/alignment.py]:
```suggestion
                    "Indexes along dimension %r don't have the same length."
```

File: xarray/core/alignment.py (hunk 5 above)
Comment (dcherian) [reply xarray/core/alignment.py]: I like that.

**Comment by shoyer** [comment]
LGTM -- anyone else have opinions here?

**Comment by max-sixty** [comment]
Yes agree! Thanks @dcherian !

**Comment by dcherian** [comment]
Great, thanks!

**Pull request merged**